ipyopenlayers/Map.py

```diff
@@ -166,4 +170,73 @@ def remove_control(self, control):
     
 
     def clear_layers(self):
-        self.layers = []
\ No newline at end of file
+        self.layers = []
+
+    def on_msg(self, callback):
+        """Register a callback for receiving messages from the frontend."""
+        self._msg_callback = callback
+
+    def _handle_msg(self, msg):
+        """Handle a message received from the frontend."""
+        if hasattr(self, '_msg_callback'):
+            content = msg.get('content', {})
+            buffers = msg.get('buffers', [])
+            self._msg_callback(self, content, buffers)
+            
+    def _handle_click(self, lat, lon, api_key):
+        """
+        Handle click events and trigger registered callbacks.
+        """
+        print(f"Handling click event at Longitude: {lon}, Latitude: {lat}")
+        country = self.get_country_from_coordinates_geoapify(lat, lon, api_key)
+        self.clicked_country = country
+        print(f"Country: {country}")
+
+        callbacks = self._click_callbacks.get_callbacks() if self._click_callbacks else []
+        for callback in callbacks:
+            if callable(callback):
+                callback(lat, lon, country)
+
+    def on_click(self, callback, api_key):
+        """
+        Register a callback to handle click events on the map and pass the API key.
+        """
+        self._click_callback = callback
+        
+        def handle_frontend_event(widget, content, buffers):
+            """Handle the click event from the frontend."""
+            data = content.get('data', {})
+            method = data.get('method', '')
+
+            if method == 'custom':
+                event_data = data.get('content', {})
+                lon = event_data.get('lon')
+                lat = event_data.get('lat')
+                
+                if callable(self._click_callback):
+                    self._click_callback(lat, lon, self.get_country_from_coordinates_geoapify(lat, lon, api_key))
```

Comment:
Do we really need to create a new `_click_callback` property here?

```suggestion
        def handle_frontend_event(widget, content, buffers):
            """Handle the click event from the frontend."""
            data = content.get('data', {})
            method = data.get('method', '')

            if method == 'custom':
                event_data = data.get('content', {})
                lon = event_data.get('lon')
                lat = event_data.get('lat')
                
                callback(lat, lon, self.get_country_from_coordinates_geoapify(lat, lon, api_key))
```